capstone/src/test/java/com/google/sps/servlets/ChallengeServletTest.java

```diff
@@ -0,0 +1,168 @@
+    responseWriter = new StringWriter();
+    when(mockResponse.getWriter()).thenReturn(new PrintWriter(responseWriter));
+
+    challengeServlet = new ChallengeServlet();
+  }
+
+  private Entity createChallenge(String text) {
+    Entity challengeEntity = new Entity("Challenge");
+    challengeEntity.setProperty("name", text);
+    challengeEntity.setProperty("dueDate", this.getDueDate());
+    challengeEntity.setProperty("votes", new ArrayList<String>());
+    challengeEntity.setProperty("timestamp", System.currentTimeMillis());
+    return challengeEntity;
+  }
+
+  /**
+   * Sets due date to midnight, 7 days from when challenge is posted.
+   *
+   * @return due date in milliseconds
+   */
+  private long getDueDate() {
+    LocalDateTime now = LocalDateTime.now();
+    LocalDateTime dueDate = now.plusDays(7).withHour(23).withMinute(59).withSecond(59).withNano(0);
+    return Timestamp.valueOf(dueDate).getTime();
+  }
+
+  @After
+  public void tearDown() {
+    helper.tearDown();
+    responseWriter = null;
+    datastore = null;
+    challengeServlet = null;
+  }
+
+  @Test
+  public void doGet_testChallengeName() throws IOException {
+    challengeServlet.doGet(mockRequest, mockResponse);
+    String response = responseWriter.toString();
+    assertTrue(response.contains(CHALLENGE_NAME));
+  }
+
+  @Test
+  public void doGet_testUserNotCompleted() throws IOException {
+    challengeServlet.doGet(mockRequest, mockResponse);
+    String response = responseWriter.toString();
+    assertTrue(response.contains("false"));
+  }
+
+  @Test
+  public void doGet_userNotLoggedIn() throws IOException {
+    helper.setEnvIsLoggedIn(false);
+
+    challengeServlet.doGet(mockRequest, mockResponse);
+    String response = responseWriter.toString();
+
+    assertTrue(response.contains("Oops an error happened!"));
+  }
+
+  @Test
+  public void doPost_userLoggedIn() throws IOException, EntityNotFoundException {
+    when(mockRequest.getParameter("name")).thenReturn(NEW_CHALLENGE);
+
+    challengeServlet.doPost(mockRequest, mockResponse);
+    // id of Challenge increments for each new Challenge that is added
+    Key challengeKey = KeyFactory.createKey("Challenge", CHALLENGE_ID);
+    Entity entity = datastore.get(challengeKey);
+    long id = entity.getKey().getId();
+    long dueDate = this.getDueDate();
+    //    long dueDateMillis = Timestamp.valueOf(dueDate).getTime();
+    Challenge challenge =
+        new Challenge(
+            NEW_CHALLENGE, /* challenge name*/
+            dueDate, /* due date */
+            null, /* badge */
+            new ArrayList<String>(), /* users completed */
+            id /* id of challenge */);
+    Challenge returnedChallenge = Challenge.fromEntity(entity);
+
+    assertEquals(challenge.getChallengeName(), returnedChallenge.getChallengeName());
+    assert challenge.getUsersCompleted().size() == returnedChallenge.getUsersCompleted().size();
+    assert challenge.getDueDate() == returnedChallenge.getDueDate();
```

Comment:
You could override the challenge .equals() method to compare these instead of manually comparing each field. Or you could convert both challenge objects to JSON strings and compare the JSON.